dsa.go

```diff
@@ -93,55 +93,47 @@ func (c *Context) GenerateDSAKeyPairWithLabel(id, label []byte, params *dsa.Para
 		return nil, errClosed
 	}
 
-	if err := notNilBytes(id, "id"); err != nil {
-		return nil, err
-	}
-	if err := notNilBytes(label, "label"); err != nil {
+	template, err := NewAttributewithLabel(id, label)
+	if err != nil {
 		return nil, err
 	}
-
-	return c.generateDSAKeyPair(id, label, params)
+	return c.GenerateDSAKeyPairWithAttributes(template, template.Copy(), params)
 }
 
-// generateDSAKeyPair creates a DSA private key on the token.
-func (c *Context) generateDSAKeyPair(id, label []byte, params *dsa.Parameters) (k *pkcs11PrivateKeyDSA, err error) {
+// GenerateDSAKeyPairWithAttributes creates a DSA key pair on the token. Additional attributes from public and
+// private will be merged into the default templates used when generating keys. Where conflicts occur, the user
+// supplied attributes will win.
+func (c *Context) GenerateDSAKeyPairWithAttributes(public, private *AttributeSet, params *dsa.Parameters) (k *pkcs11PrivateKeyDSA, err error) {
+	if err = validateKeyPairAttributes(public, private); err != nil {
+		return
+	}
+
 	err = c.withSession(func(session *pkcs11Session) error {
 		p := params.P.Bytes()
 		q := params.Q.Bytes()
 		g := params.G.Bytes()
 
-		publicKeyTemplate := []*pkcs11.Attribute{
+		public.AddDefaultAttributes([]*pkcs11.Attribute{
```

Comment:
It might be polite to take a copy of these templates before we mutate them. Or... we see this as a feature and explain in the function comments that the final applied template is "returned" in the calling arguments. What do you think?

(This comment applies to all the `...WithAttributes` functions).

Reply:
Sorry, I've just remembered your introduction:

>Removed the code that returned the modified Attribute slices since the new AttributeSet is mutable and will contain the added Attributes after the call.

So that implies we will comment this as a feature, which I'm happy with :-)

Reply:
The original commit was trying to return the modified AttributeSets. I felt this was very hard to maintain so I decided it would be better to just update the received AttributeSet so the caller can manipulate for subsequent calls if they needed to. 

I can make a clone instead and just not return the modified `AttributeSet`. What method would you prefer?